### ISSUE TITLE
pkg.exports field in bad order for webpack@5

### ISSUE THREAD

**Issue by ctavan** [ai/dual-publish#10]
While reading https://github.com/facebook/jest/issues/9430#issuecomment-638489222 I realized that the keys in the new `pkg.exports` field are probably in the wrong order. `dual-publish` currently seems to produce:

```json
".": {
  "require": "./index.cjs",
  "import": "./index.js",
  "browser": "./index.browser.js"
},
```

(at least that's in https://cdn.jsdelivr.net/npm/nanoid@3.1.9/package.json).

I've learnt in https://github.com/uuidjs/uuid/pull/462#issuecomment-637571626 that object key order matters here!

I assume the intention of the above exports block is that when bundling for browsers, bundlers should always pick `index.browser.js`? As far as I understand the way `webpack@5` implements this this would not be the case with the above exports block!

Maybe @evilebottnawi or @sokra can confirm this.

/cc @jkrems and @guybedford to raise awareness around how `pkg.exports` is currently understood and implemented by the community.

**Comment by evilebottnawi** [comment]
Resolution order follows keys order in exports field. 

> Condition matching is applied in object order from first to last within the "exports" object. The general rule is that conditions should be used from most specific to least specific in object order.

```
".": {
  "require": "./index.cjs",
  "browser": "./index.browser.js"
  "import": "./index.js",
},
```

But it will pick up the `browser` field only for `target: web`.

**Comment by sokra** [comment]
``` js
​".": {
  "browser": "./index.browser.js"
  "require": "./index.cjs",
  "default": "./index.js",
},
```

**Comment by ai** [comment]
Thanks for the notice. I will update the code and publish fix for all my libraries on next week (sorry, on this weekend I am in outdoor vacation).

**Comment by ctavan** [comment]
> ```js
> ​".": {
>   "browser": "./index.browser.js"
>   "require": "./index.cjs",
>   "default": "./index.js",
> },
> ```

@sokra @evilebottnawi @jkrems @guybedford: Is it generally considered a „best practice“ to specify the esm entrypoint as „default“ instead of „import“?

**Comment by jkrems** [comment]
You may get different answers here. I would personally say “yes”. Others feel that using import is more explicit and safer. E.g. if you remove the require condition, default will make require() try to load that file and fail. Using the “import” condition, require() will fail while resolving (“no compatible export for this path”) which _may_ be more clear.

**Comment by ai** [comment]
I fixed the `exports` generation. Now its generate:

```
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "browser": "./index.browser.js",
      "require": "./index.cjs",
      "import": "./index.js",
      "default": "./index.js"
    },
```

@sokra @evilebottnawi is it correct? (especially having both `import` and `default`)

**Comment by ctavan** [comment]
Given Node.js' and webpack's implementation of `pkg.exports` the `default` case seems redundant. Since it's always harder to remove code than add code later (once tools start relying on something) I'd propose to remove the `default` case.

**Comment by ai** [comment]
@ctavan why @sokra mentioned `default`?

**Comment by jkrems** [comment]
As long as there's a `require` condition, there's effectively no difference between keeping `default`, keeping `import`, or keeping both. If you ever want to make shipping CJS optional ("click here to stop supporting CJS"), then it would become interesting.

* Keep `import`: `require('pkg')` may fail saying that `pkg` has no mapping from `require`.
* Keep `default`: `require('pkg')` resolves to `default`, then fails saying that ESM cannot be required.
* Keep both: Same as "keep `default`".

So keeping `import` would fail earlier. If you keep both, you might as well drop `import` since it's redundant (`default` is always matched).

**Comment by ctavan** [comment]
> @ctavan why @sokra mentioned `default`?

Well, [I was wondering that as well](https://github.com/ai/dual-publish/issues/10#issuecomment-639832432).

I assume the idea of specifying the ESM build as a default would be to speculate that future tools will work well with that. I personally don't like this speculative character so much and would prefer explicit specifiers (like `import`).

**Comment by sokra** [comment]
You can use `import` instead of `default` if you want to. I just wanted to handle cases that would have a different condition.

There are more ways to reference modules next to `require` and `import`. E. g. in CSS you also have `@import` and `url()`. In WASM there is something similar to `import`. There is also `new URL(import.meta.url, "...")`. These might result in other condition names next to `require` and `import`.

But it's probably better to keep them unhandled instead of defaulting them to ESM.

**Comment by sokra** [comment]
You may want to wait for the decision in this issue https://github.com/webpack/webpack/issues/11014

This may resulting in an additional `module` condition you want to handle for better optimization.

**Comment by ai** [comment]
I removed `exports[x].default` and subscribed to that issue.

Thanks. Now, let’s wait for final decision.

**Comment by ai** [comment]
I released the fix in 0.10.3. Anyway, it is just reordering.
